File: src/main/java/rars/riscv/lang/lexing/RVTokensProducer.java
package rars.riscv.lang.lexing;

import org.jetbrains.annotations.NotNull;
import rars.riscv.lang.Position;

import javax.swing.text.Segment;
import java.util.ArrayList;
import java.util.List;

public class RVTokensProducer implements TokensProducer<TokenizedLine> {
    private final Lexer<TokenizedLine, RVTokensProducer> lexer;
    private ArrayList<RVToken> result;
    private int lineNum;

    public RVTokensProducer(Lexer<TokenizedLine, RVTokensProducer> lexer) {
        this.lexer = lexer;
    }

    public RVTokensProducer() {
        this(new RVLexer<>());
    }

    @Override
    public @NotNull TokenizedLine getEmptyResult() {
        return new TokenizedLine(List.of(), lineNum);
    }

    @Override
    public TokenizedLine getTokenList(Segment text, int initialTokenType, int lineOffset, int lineNum) {
        this.result = new ArrayList<>();
        this.lineNum = lineNum;
        return lexer.getTokensList(text, initialTokenType, lineOffset, this);
    }

    @Override
    public void addToken(char @NotNull [] array, int start, int end, RVTokenType tokenType, int startOffset) {
        final var position = new Position(this.lineNum, start, startOffset);
        final var substring = new String(array, start, end - start);
        final var newToken = new RVToken(position, tokenType, substring);
        result.add(newToken);
    }

    @Override
    public @NotNull TokenizedLine getResult() {
        return new TokenizedLine(result, lineNum);
    }
}

File: src/main/java/rars/riscv/lang/lexing/RVTokensProducer.kt
package rars.riscv.lang.lexing

import rars.ErrorList
import rars.ErrorMessage
import rars.RISCVProgram
import rars.riscv.lang.Position
import javax.swing.text.Segment

class RVTokensProducer @JvmOverloads constructor(
    private val program: RISCVProgram,
    private val errorList: ErrorList,
    private val lexer: Lexer<TokenizedLine, RVTokensProducer> = RVLexer<TokenizedLine, RVTokensProducer>(),
) : TokensProducer<TokenizedLine> {
    private var lineNum = 0
    private var result = mutableListOf<RVToken>()

    override fun getEmptyResult(): TokenizedLine = TokenizedLine(emptyList(), lineNum)

    override fun getTokenList(
        text: Segment, initialTokenType: Int, lineOffset: Int,
        lineNum: Int
    ): TokenizedLine {
        this.result = ArrayList<RVToken>()
        this.lineNum = lineNum
        return lexer.getTokensList(text, initialTokenType, lineOffset, this)
    }

    override fun addToken(
        array: CharArray, start: Int, end: Int, tokenType: RVTokenType,
        startOffset: Int
    ) {
        val position = Position(lineNum, start, startOffset)
        val substring = String(array, start, end - start)
        val newToken = RVToken(position, tokenType, substring)
        result.add(newToken)
    }

    override fun addErrorToken(
        array: CharArray,
        segmentPos: Int,
        offset: Int,
        notice: String
    ) {
        this.errorList.add(ErrorMessage.error(program, lineNum, offset, notice))
        super.addErrorToken(array, segmentPos, offset, notice)
    }

    override fun getResult(): TokenizedLine = TokenizedLine(result, lineNum)
}
